src/serializer/ResidualReactElementVisitor.js

```diff
@@ -44,22 +49,35 @@ export class ResidualReactElementVisitor {
         if (!isReactFragment) {
           this.residualHeapVisitor.visitValue(typeValue);
         }
+        if (typeValue instanceof StringValue) {
+          isHostComponent = true;
+        }
       },
       visitKey: (keyValue: Value) => {
         this.residualHeapVisitor.visitValue(keyValue);
       },
       visitRef: (refValue: Value) => {
-        this.residualHeapVisitor.visitValue(refValue);
+        if (!firstRenderOnly) {
+          this.residualHeapVisitor.visitValue(refValue);
+        }
       },
       visitAbstractOrPartialProps: (propsValue: AbstractValue | ObjectValue) => {
         this.residualHeapVisitor.visitValue(propsValue);
       },
       visitConcreteProps: (propsValue: ObjectValue) => {
         for (let [propName, binding] of propsValue.properties) {
-          if (binding.descriptor !== undefined && propName !== "children") {
-            let propValue = getProperty(this.realm, propsValue, propName);
-            this.residualHeapVisitor.visitValue(propValue);
+          invariant(propName !== "key" && propName !== "ref", `"${propName}" is a reserved prop name`);
+          if (binding.descriptor === undefined || propName === "children") {
+            continue;
+          }
+          let propValue = getProperty(this.realm, propsValue, propName);
+
+          // In firstRenderOnly mode, we strip off onEventHanlders and any props
+          // that are functions as they are not required for init render.
+          if (firstRenderOnly && isHostComponent && (isEventProp(propName) || propValue instanceof FunctionValue)) {
```

Comment:
Can we make this a shared helper that takes all these things as arguments? So it’s obvious this must be kept in sync with identical logic in the other file.

Reply:
Good idea.